metagamer/environments/qagent.py

```diff
@@ -1,144 +1,432 @@
+    def get_action(self, state, valid_actions):
+        return random.choice(valid_actions)
+
+
+class PageLines(Agent):
+    def get_action(self, state, valid_actions):
+        """
+        return a possible position that hasn't yet been played.
+        This policy tries simply to fill the board from the top left.
+        Args:
+            state: a 3*3 numpy array whose elements are 0, 1, -1 representing empty, player1 and player 2,
+            valid_actions: either 1 or -1, representing player 1 and player 2
+
+
+        Returns: a tuple representing an unfilled coordinate that's selected.
+
+        """
+        for row in range(state.shape[0]):
+            for col in range(state.shape[1]):
+                if state[row, col] == 0:
+                    return row, col
+
+
+class EpsilonPageLines(Agent):
+    def get_arg_max(self, state, valid_actions):
+        """
+        return a possible position that hasn't yet been played.
+        This policy tries simply to fill the board from the top left.
+        Args:
+            board: a 3*3 numpy array whose elements are 0, 1, -1 representing empty, player1 and player 2,
+            player: either 1 or -1, representing player 1 and player 2
+
+
+        Returns: a tuple representing an unfilled coordinate that's selected.
+
+        """
+        for row in range(state.shape[0]):
+            for col in range(state.shape[1]):
+                if state[row, col] == 0:
+                    return row, col
+
+    # qtable
+    # random
+    # dqn
+    # policy gradient
+    # left to right
+
+
+class QTicTacTable(Agent):
+    def wrapper(self, env):
+        return QtableWrapper(env)
+
+    def __init__(self, alpha=0.1, gamma=0.9):
+        """
+        this will contain the state
+        qdict[state][action] = qvalue
+        where state is a string representing the board and action is an integer
+        """
+        # choice to prune or build
+        # we choose prune here since it's less likely to go wrong
+        super().__init__()
+        self.qdict: Dict[str, Dict[int, float]] = collections.defaultdict(
+            lambda: {i: 0 for i in range(9)}
+        )
+        # update decay - smaller alphas have longer memory, larger alphas shorter
+        self.alpha = alpha
+        # discount factor when updating reward - high gamma is more greedy, low gamma is more patient
+        self.discount = gamma
 
-    def train(self, num_epsiodes, initeps=1, finaleps=0.05):
+    def set_reward(self, state: str, action: int, value: float):
         """
-        Simple linear drop for epsilon.
+        Only allowed to set a state, value pair.
+        Update is as per alpha value
         """
-        epsdecay = (initeps - finaleps) / num_epsiodes
-        epsilon = initeps
+        if self.eval:
+            # no learning in eval mode
+            return
+        self.qdict[state][action] = (
+            self.qdict[state][action] * (1 - self.alpha) + value * self.alpha
+        )
```

Comment:
So alpha determines how much the most recent reward overwrites past memory of rewards in that state. Seems good.